### PR TITLE
adding linear regression component and rule component

### PR DESCRIPTION
**Pull request by cindyheqy**

### CODE REVIEW

File: dextra-components/src/components/Marks/Regression/src/Regression.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { customElement, property } from "lit/decorators.js";
+import { PlotMark } from "../../PlotMark/src/PlotMark";
+import * as Plot from "@observablehq/plot";
+import { LinearRegressionOptions } from "@observablehq/plot";
+import { LinearRegressionYOptions } from "@observablehq/plot";
+
+
+@customElement("osl-regression-plot")
+export class OslRegressionPlot extends PlotMark {
+  mark = "regression"
```
Comment (nofurtherinformation) [inline dextra-components/src/components/Marks/Regression/src/Regression.ts:10]: Repo typescript settings have changed slightly and the repo requires more explicit overrides. For class methods or properties that overwrite the base class, you should use the keyword `override` such as `override mark` - you should see this flagged in the VS Code type hints/errors now.

**Comment by nofurtherinformation** [comment]
@cindyheqy I just added some initial thoughts. You've done great work. Let me know if you have any questions on my review!

File: dextra-components/src/components/Marks/Regression/src/Regression.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nofurtherinformation) [reply dextra-components/src/components/Marks/Regression/src/Regression.ts:10]: A small update here -- due to some build tool conflicts, no need to use override any more, ignore this previous comment!

File: dextra-components/src/components/Marks/RuleX/src/RuleX.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  @property({ type: String })
+  y2?: RuleXOptions["y2"];
+
+  @property({ type: Array })
+  number?: [0];  
+
+  override markOptions: (keyof this)[] = ["interval", "x", "y", "y1", "y2", "number"];
+
+  public override get plot() {
```
[the remark this answers is not in the document]
Comment (cindyheqy) [reply dextra-components/src/components/Marks/RuleX/src/RuleX.ts]: ok. combine the two into Rule component with boolean property 'vertical' switching between Rule X and Rule Y

File: dextra-components/src/components/Marks/RuleX/src/RuleX.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    const innerData = this?.currentResults;
+    const options = this.allOptions;
+
+    return (overrideData?: any, overrideOptions?: {}) => {
+      const data = innerData?.length ? innerData : overrideData;
+      if (this.number) {
```
[the remark this answers is not in the document]
Comment (cindyheqy) [reply dextra-components/src/components/Marks/RuleX/src/RuleX.ts]: changed data type to ‘converter: interpretFuncJsonOrStrin’ and deleted 'number' property

File: dextra-components/src/components/Marks/RuleX/src/RuleX.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  y?: RuleXOptions["y"];
+
+  @property({ type: String })
+  y1?: RuleXOptions["y1"];
+
+  @property({ type: String })
+  y2?: RuleXOptions["y2"];
+
+  @property({ type: Array })
+  number?: [0];
```
[the remark this answers is not in the document]
Comment (cindyheqy) [reply dextra-components/src/components/Marks/RuleX/src/RuleX.ts]: changed data type to ‘converter: interpretFuncJsonOrStrin’ and deleted 'number' property

File: dextra-components/src/components/Marks/Regression/src/Regression.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  markOptions: (keyof this)[] = ["x", "y", "ci"];
+
+  public get plot() {
+    const innerData = this?.currentResults;
+    const options = this.allOptions;
+
+    return (overrideData?: any, overrideOptions?: {}) => {
+      const data = innerData?.length ? innerData : overrideData;
+      return Plot.linearRegressionY(data, { ...overrideOptions, ...options});
```
[the remark this answers is not in the document]
Comment (cindyheqy) [reply dextra-components/src/components/Marks/Regression/src/Regression.ts]: Added a property 'regressionX' that takes in boolean value to switch between the default 'regressionY' and regressionX'

**Comment by nofurtherinformation** [comment]
@cindyheqy This looks good, thanks for your work!

A few small things:
- For Rule, instead of `vertical: bool` let's use `direction: 'horizontal' | 'vertical'` - I think that is a more intuitive API
- Can you add an example of rule being used to show bounds like [this example](https://observablehq.com/@observablehq/observable-plot-candlestick)? Here is a [sample data endpoint with min/max fields that should work for that data shape](https://d3uldu0bz6pkei.cloudfront.net/data-query/647e4573a0fa480008e0fbbd)

**Comment by cindyheqy** [comment]
d7f4786ac1802ad91af335fc4a1a6bf3a5639790 resolve the date error

**Review by nofurtherinformation — APPROVED** [review]

**Comment by nofurtherinformation** [comment]
Approved to merge!

**Pull request merged**